src/main/java/org/javarosa/core/model/actions/SetValueAction.java

```diff
@@ -68,43 +68,24 @@ public void handle(XFormParser p, Element e, Object parent) {
     }
 
     public TreeReference processAction(FormDef model, TreeReference contextRef) {
-
-        //Qualify the reference if necessary
+        // If the target is in a repeat, it's stored as an unbound ref. Use the context that the action is defined in to qualify it
         TreeReference targetReference = contextRef == null ? target : target.contextualize(contextRef);
-
-        //For now we only process setValue actions which are within the
-        //context if a context is provided. This happens for repeats where
-        //insert events should only trigger on the right nodes
-        if (contextRef != null){
-            //Note: right now we're qualifying then testing parentage to see whether
-            //there was a conflict, but it's not super clear whether this is a perfect
-            //strategy
-            if (!contextRef.isAncestorOf(targetReference, false)) {
-                return null;
-            }
-        }
-
-        //TODO: either the target or the value's node might not exist here, catch and throw reasonably
         EvaluationContext context = new EvaluationContext(model.getEvaluationContext(), targetReference);
 
         String failMessage = "Target of TreeReference " + target.toString(true) + " could not be resolved!";
 
-        if (targetReference.hasPredicates()) {
-            //CTS: in theory these predicates could contain logic which breaks if the qualified ref
-            //contains unbound repeats (IE: nested repeats).
-            List<TreeReference> references = context.expandReference(targetReference);
-            if (references.size() == 0) {
-                //If after finding our concrete reference it is a template, this action is outside of the
-                //scope of the current target, so we can leave.
-                if (model.getMainInstance().hasTemplatePath(target)) {
-                    return null;
-                }
-                throw new NullPointerException(failMessage);
-            } else if (references.size() > 1) {
-                throw new XPathTypeMismatchException("XPath nodeset has more than one node [" + references + "]; Actions can only target a single node reference. Refine path expression to match only one node.");
-            } else {
-                targetReference = references.get(0);
+        List<TreeReference> references = context.expandReference(targetReference);
+        if (references.size() == 0) {
+            // If after finding our concrete reference it is a template, this action is outside of the
+            // scope of the current target, so we can leave.
+            if (model.getMainInstance().hasTemplatePath(target)) {
```

Comment:
I have been unable to come up with a form that would result in either an expanded reference set of size 0 or a reference with a template path. It feels safer to leave these cases in for now, though.

Reply:
For 0 refs, I believe FormInstanceParser.verifyActions makes that impossible. 

TODO: try a predicate that does lead to a node being found at parse time but not at event time.

Reply:
I can get hit the `references.size() == 0` and then `hasTemplatePath` code path here if I **comment out the contents of `verifyActions`** and run the following:

```
   @Test
    public void foo() throws IOException {
        Scenario scenario = Scenario.init("Setvalue into repeat", html(
            head(
                title("Setvalue into repeat"),
                model(
                    mainInstance(t("data id=\"setvalue-into-repeat\"",
                        t("dest_index", "1"),
                        t("source"),
                        t("repeat",
                            t("destination")
                        )
                    ))
                )
            ),
            body(
                input("/data/dest_index"),
                input("/data/source",
                    setvalue("xforms-value-changed", "/data/repeat[position()=/data/dest_index]/destination", "/data/source")),
                repeat("/data/repeat",
                    input("/data/repeat/destination")
                ))));

        scenario.createNewRepeat("/data/repeat");
        scenario.createNewRepeat("/data/repeat");
        scenario.createNewRepeat("/data/repeat");

        scenario.answer("/data/source", "foo");
        assertThat(scenario.answerOf("/data/repeat[0]/destination").getDisplayText(), is("foo"));

        scenario.answer("/data/dest_index", "7");
        scenario.answer("/data/source", "bar");
    }
```

Again, I think `verifyActions` is doing more than it should be, but it makes sense to me to wait for a user to report.